module/zfs/sa.c

```diff
@@ -2010,7 +2167,25 @@ sa_handle_unlock(sa_handle_t *hdl)
 	mutex_exit(&hdl->sa_lock);
 }
 
+sa_handle_t *
+sa_handle_alloc(objset_t *os, dmu_buf_t *db, void *userp)
```

Comment:
ah, my bad - I forgot to remove this from the patch. in short, the problem is that to be able to update attributes (even regular), we need a SA handle. sa_handle_get() builds the index which starts with an empty EA, then we have to add EA(s) in all the cases, so we rebuild the index again. and each rebuild has to find an appropriate layout, then check for an existing index table.. which is quite expensive. so the idea was to get an empty SA handle and build the layout/index when we instantiate attributes along with mandatory EAs.
what would be your idea to this problem?